lib/ddtrace/profiling/ext/thread.rb

```diff
@@ -0,0 +1,75 @@
+require 'ffi'
+
+module Datadog
+  module Profiling
+    module Ext
+      # C-struct for retrieving clock ID from pthread
+      class CClockId < FFI::Struct
+        layout :value, :int
+      end
+
+      # Extensions for pthread-backed Ruby threads, to retrieve
+      # the thread ID, clock ID, and CPU time.
+      module CThread
+        extend FFI::Library
+        ffi_lib 'ruby', 'pthread'
+        attach_function :rb_nativethread_self, [], :ulong
+        attach_function :pthread_getcpuclockid, [:ulong, CClockId], :int
+
+        def self.prepended(base)
+          # Be sure to update the current thread too; as it won't have been set.
+          unless Thread.current.native_thread_id
+            Thread.current.instance_variable_set(:@native_thread_id, CThread.rb_nativethread_self)
+          end
+        end
+
+        attr_reader :native_thread_id
+
+        def initialize(*args)
+          # Wrap the work block with our own
+          # so we can retrieve the native thread ID within the thread's context.
+          wrapped_block = proc do |*t_args|
+            @native_thread_id = rb_nativethread_self
+            yield(*t_args)
+          end
+
+          super(*args, &wrapped_block)
+        end
+
+        def clock_id
+          return unless native_thread_id
+
+          # Build a struct, pass it to Pthread's getcpuclockid function.
+          clock = CClockId.new
+          clock[:value] = 0
+          return_code = pthread_getcpuclockid(native_thread_id, clock)
```

Comment:
I'm not sure where/when this function is going to be called, but be ultra-careful: if you call this on a `native_thread_id` that is invalid (e.g. thread finished), this will *crash* Ruby (segfault).

Reply:
Hmmm...

The native thread ID is resolved and stored on the `Thread` when the thread begins. This function is invoked later by the profiler when its time to get the CPU. I suppose it's possible that between when the profiler evaluates `Thread.list`, and this function is invoked, it's possible that the thread finishes and maybe this becomes invalid? I think you might be right about the risk.

Maybe instead I can move this into the thread start block as well, then cache the clock ID. This adds a slight overhead to the thread, but might be worthwhile if it avoids segfaults. Of course, this depends on `Process.get_clocktime` not segfaulting either when you give it an obsolete clock (I don't think it does.)

Reply:
Moved clock resolution into the thread that owns it; it shouldn't be possible for a segfault because the thread ID has become invalid after finishing now, because its always evaluated from within the thread that owns it (thread must be running to do this evaluation.)

Should be resolved?

Reply:
Yeah that should be fine this way!

Reply:
Okay one weird scenario with doing this.

When Ruby forks, the clock ID stored on the thread is no longer valid within the fork (`cpu_time` raises an error.) This meant in Puma, for example, when a web worker would fork, it would raise an exception while retrieving `cpu_time` for the main thread within the fork.

So we have to update the clock ID I think? Issue with doing this is that it puts us back where we started; you have no guarantee the pthread is running, and its technically possible the pthread terminates between the time the thread is retrieved via `Thread.list` and when `cpu_time` is evaluated/clock is updated for that thread.

I tried to address this with https://github.com/DataDog/dd-trace-rb/pull/1097/commits/f405a567afa15c1a44454e20d3edc9ffbba609b6 but I don't know if this is sufficient. Thoughts @jd ?

Reply:
Yes, I would definitely make the main thread a special case. It actually has its own class/singleton in Python threading module, because it *is* special. ;)

Your workaround is good though if you prefer that approach.